File: DSA/DSA07021 DÃY NGOẶC ĐÚNG DÀI NHẤT.cpp
#include <bits/stdc++.h>
using namespace std;

void submit()
{
    string s;
    getline(cin, s);

    stack<int> A;
    A.push(-1);
    int maxi = 0;
    for (int i = 0; i < s.size(); i++)
    {
        char c = s[i];
        if (c == '(' || c == '[' || c == '{')
        {
            A.push(i);
        }
        else
        {
            if ((!A.empty() && A.top() != -1 &&
               ((c == ')' && s[A.top()] == '(')||
                (c == ']' && s[A.top()] == '[')||
                (c == '}' && s[A.top()] == '{'))))
            {
                A.pop();
                maxi = max(maxi, i-A.top());
            }
            else
            {
                A.push(i);
            }
        }
    }
    
    
    cout << maxi;

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA07021 DÃY NGOẶC ĐÚNG DÀI NHẤT.py
def submit():
    s = input()

    maxi = 0
    A = list()
    A.append(-1)

    for i in range(len(s)):
        c = s[i]
        if c == '(' or c == '[' or c == '{':
            A.append(i)
        else:
            if (len(A) != 0 and A[-1] != -1 and
                (c == ')' and s[A[-1]] == '(') or
                (c == ']' and s[A[-1]] == '[') or
                (c == '}' and s[A[-1]] == '{')):
                A.pop()
                maxi = max(maxi, i-A[-1])
            else:
                A.append(i)


    print(maxi)




test = int(input())
for i in range(test):
    submit()